{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bd7d3448-7cd2-476a-991f-543a571e152a",
   "metadata": {},
   "source": [
    "## V0_4 Updates\n",
    "1. Change Perturbation Encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4b4a182-be86-4c90-93cf-8008d5f38e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "from tqdm import tqdm\n",
    "from pathlib import Path\n",
    "from gears import PertData\n",
    "from pathlib import Path\n",
    "from Bio import Entrez, SeqIO\n",
    "import mygene\n",
    "import gzip\n",
    "from transformers import AutoTokenizer, AutoModel, AutoModelForMaskedLM\n",
    "import torch\n",
    "from sklearn.model_selection import train_test_split\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bfe00742-3648-45b2-ac04-b8b203574b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path('/home/ubuntu')\n",
    "tok_dir = data_dir / 'training'\n",
    "raw_dir = data_dir / 'raw_data'\n",
    "protein_dir = data_dir / 'uniprot'\n",
    "pert_dir = data_dir / 'pert_embd'\n",
    "\n",
    "splits= ['train', 'val', 'test']\n",
    "\n",
    "crispri_seq = raw_dir / 'crispri_seq.csv'\n",
    "dataset_name = 'k562e'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "75b03165-2beb-4c31-bf32-85d101560db3",
   "metadata": {},
   "outputs": [],
   "source": [
    "chunk_size = 20000        # How many cells per file\n",
    "n_genes = 8192 # 2**13\n",
    "count_normalize_target = 1e4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "df468668-d8ae-4df4-ba71-71db2d18f036",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "using cuda\n"
     ]
    }
   ],
   "source": [
    "def get_device():\n",
    "    device = 'cpu'\n",
    "    if torch.cuda.is_available():\n",
    "        torch.cuda.manual_seed(1337)\n",
    "        device = 'cuda'\n",
    "    print(f'using {device}')\n",
    "    return device\n",
    "\n",
    "device = get_device()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ebfb1985-fdbd-469e-9a80-b80939d9a626",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants for BioJEPA\n",
    "modality_to_id = {\n",
    "    'protein': 0,\n",
    "    'chemical': 1,\n",
    "    'dna': 2\n",
    "}\n",
    "\n",
    "mode_to_id = {\n",
    "    'crispri': 0,\n",
    "    'crispra': 1,\n",
    "    'overexpression': 2,\n",
    "    'knockout': 3,\n",
    "    'inhibitor': 4,\n",
    "    'agonist': 5,\n",
    "    'degrader': 6,\n",
    "    'binder': 7,\n",
    "    'control': 8,\n",
    "    'unknown': 9\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6fa2532-aa8a-4cdf-b9d0-c055e5481e73",
   "metadata": {},
   "source": [
    "## Data Download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c03df92b-cd7b-43c8-8ba9-dc82f4b3f411",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading...\n",
      "100%|███████████████████████████████████████████████████████████████| 9.46M/9.46M [00:00<00:00, 120MiB/s]\n",
      "Downloading...\n",
      "100%|████████████████████████████████████████████████████████████████| 670M/670M [00:25<00:00, 26.4MiB/s]\n",
      "Extracting zip file...\n",
      "Done!\n",
      "Downloading...\n",
      "100%|████████████████████████████████████████████████████████████████| 559k/559k [00:00<00:00, 23.7MiB/s]\n",
      "These perturbations are not in the GO graph and their perturbation can thus not be predicted\n",
      "['C7orf26+ctrl' 'C14orf178+ctrl' 'RPS10-NUDT3+ctrl' 'SEM1+ctrl' 'FAU+ctrl']\n",
      "Creating pyg object for each cell in the data...\n",
      "Creating dataset file...\n",
      "100%|████████████████████████████████████████████████████████████████| 1088/1088 [07:56<00:00,  2.28it/s]\n",
      "Done!\n",
      "Saving new dataset pyg object at /home/ubuntu/k562e/replogle_k562_essential/data_pyg/cell_graphs.pkl\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "pert_data = PertData(data_dir / dataset_name) \n",
    "pert_data.load(data_name='replogle_k562_essential')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b065529f-7e52-4946-9355-9108d93b968c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating new splits....\n",
      "Saving new splits at /home/ubuntu/k562e/replogle_k562_essential/splits/replogle_k562_essential_simulation_1_0.75.pkl\n",
      "Simulation split test composition:\n",
      "combo_seen0:0\n",
      "combo_seen1:0\n",
      "combo_seen2:0\n",
      "unseen_single:272\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "pert_data.prepare_split(split='simulation', seed=1) \n",
    "adata = pert_data.adata "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "610c71db-43eb-4080-a42b-6f6d9c214cb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cov_drug_dose_name</th>\n",
       "      <th>dose_val</th>\n",
       "      <th>control</th>\n",
       "      <th>condition_name</th>\n",
       "      <th>split</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGCCAC-34</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATTCTCTCGAC-42</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACGGGAGTAATGATG-25</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGAACAAGCTAGATA-35</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGACTCTCTATTGTC-33</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTATTGCCACGTGAGA-26</th>\n",
       "      <td>RPS2+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTGTTCATGGATC-47</th>\n",
       "      <td>RPS2+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCACATCTCTTCAA-41</th>\n",
       "      <td>RPS2+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCAGTGTAGAGTTA-18</th>\n",
       "      <td>RPS2+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCCTCTCGTTCGCT-42</th>\n",
       "      <td>RPS2+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_RPS2+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>162246 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     condition cell_type  cov_drug_dose_name dose_val control  \\\n",
       "cell_barcode                                                                    \n",
       "AAACCCAAGAAGCCAC-34  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AAAGGATTCTCTCGAC-42  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AACGGGAGTAATGATG-25  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AAGAACAAGCTAGATA-35  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AAGACTCTCTATTGTC-33  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "...                        ...       ...                 ...      ...     ...   \n",
       "TTATTGCCACGTGAGA-26  RPS2+ctrl      K562  K562_RPS2+ctrl_1+1      1+1       0   \n",
       "TTGTTGTTCATGGATC-47  RPS2+ctrl      K562  K562_RPS2+ctrl_1+1      1+1       0   \n",
       "TTTCACATCTCTTCAA-41  RPS2+ctrl      K562  K562_RPS2+ctrl_1+1      1+1       0   \n",
       "TTTCAGTGTAGAGTTA-18  RPS2+ctrl      K562  K562_RPS2+ctrl_1+1      1+1       0   \n",
       "TTTCCTCTCGTTCGCT-42  RPS2+ctrl      K562  K562_RPS2+ctrl_1+1      1+1       0   \n",
       "\n",
       "                         condition_name  split  \n",
       "cell_barcode                                    \n",
       "AAACCCAAGAAGCCAC-34  K562_UBL5+ctrl_1+1  train  \n",
       "AAAGGATTCTCTCGAC-42  K562_UBL5+ctrl_1+1  train  \n",
       "AACGGGAGTAATGATG-25  K562_UBL5+ctrl_1+1  train  \n",
       "AAGAACAAGCTAGATA-35  K562_UBL5+ctrl_1+1  train  \n",
       "AAGACTCTCTATTGTC-33  K562_UBL5+ctrl_1+1  train  \n",
       "...                                 ...    ...  \n",
       "TTATTGCCACGTGAGA-26  K562_RPS2+ctrl_1+1  train  \n",
       "TTGTTGTTCATGGATC-47  K562_RPS2+ctrl_1+1  train  \n",
       "TTTCACATCTCTTCAA-41  K562_RPS2+ctrl_1+1  train  \n",
       "TTTCAGTGTAGAGTTA-18  K562_RPS2+ctrl_1+1  train  \n",
       "TTTCCTCTCGTTCGCT-42  K562_RPS2+ctrl_1+1  train  \n",
       "\n",
       "[162246 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e493a84-7723-48d1-8c2a-3761aeba848b",
   "metadata": {},
   "source": [
    "## Add in sgRNA sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "146dd3b4-55fb-471f-9347-c4043ac54660",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gem_group</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>UMI_count</th>\n",
       "      <th>sgID_AB</th>\n",
       "      <th>guide_seq_a</th>\n",
       "      <th>guide_seq_b</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAATCCA-27</th>\n",
       "      <td>NAF1</td>\n",
       "      <td>27</td>\n",
       "      <td>ENSG00000145414</td>\n",
       "      <td>11438.0</td>\n",
       "      <td>NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2</td>\n",
       "      <td>GGAGCCGTGAGCTTGTCCAG</td>\n",
       "      <td>GCCGCGACGGCGTTCAGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACTTCC-31</th>\n",
       "      <td>BUB1</td>\n",
       "      <td>31</td>\n",
       "      <td>ENSG00000169679</td>\n",
       "      <td>5342.0</td>\n",
       "      <td>BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2</td>\n",
       "      <td>GGACAAGCGCCGGGCCTCAG</td>\n",
       "      <td>GCGGGCCTCAGCGGAACCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGCCAC-34</th>\n",
       "      <td>UBL5</td>\n",
       "      <td>34</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>17305.0</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>GGGTGAGGAGCTGGTGGCGT</td>\n",
       "      <td>GCCCAGGGCCGCGAACCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAATAGTC-43</th>\n",
       "      <td>C9orf16</td>\n",
       "      <td>43</td>\n",
       "      <td>ENSG00000171159</td>\n",
       "      <td>30244.0</td>\n",
       "      <td>C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...</td>\n",
       "      <td>GGCCGGCGCCGGATGGAAGG</td>\n",
       "      <td>GGCCGCGCGACGATGGAACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGACAGCGT-28</th>\n",
       "      <td>TIMM9</td>\n",
       "      <td>28</td>\n",
       "      <td>ENSG00000100575</td>\n",
       "      <td>8407.0</td>\n",
       "      <td>TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2</td>\n",
       "      <td>GGGGACGGTTGAGCCTTGGG</td>\n",
       "      <td>GGGTTGAGCCTTGGGAGGGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene  gem_group          gene_id  UMI_count  \\\n",
       "cell_barcode                                                          \n",
       "AAACCCAAGAAATCCA-27     NAF1         27  ENSG00000145414    11438.0   \n",
       "AAACCCAAGAACTTCC-31     BUB1         31  ENSG00000169679     5342.0   \n",
       "AAACCCAAGAAGCCAC-34     UBL5         34  ENSG00000198258    17305.0   \n",
       "AAACCCAAGAATAGTC-43  C9orf16         43  ENSG00000171159    30244.0   \n",
       "AAACCCAAGACAGCGT-28    TIMM9         28  ENSG00000100575     8407.0   \n",
       "\n",
       "                                                               sgID_AB  \\\n",
       "cell_barcode                                                             \n",
       "AAACCCAAGAAATCCA-27  NAF1_+_164087918.23-P1P2|NAF1_-_164087674.23-P1P2   \n",
       "AAACCCAAGAACTTCC-31  BUB1_-_111435363.23-P1P2|BUB1_-_111435372.23-P1P2   \n",
       "AAACCCAAGAAGCCAC-34      UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AAACCCAAGAATAGTC-43  C9orf16_+_130922603.23-P1P2|C9orf16_+_13092264...   \n",
       "AAACCCAAGACAGCGT-28  TIMM9_-_58893843.23-P1P2|TIMM9_-_58893848.23-P1P2   \n",
       "\n",
       "                              guide_seq_a           guide_seq_b  \n",
       "cell_barcode                                                     \n",
       "AAACCCAAGAAATCCA-27  GGAGCCGTGAGCTTGTCCAG  GCCGCGACGGCGTTCAGAAC  \n",
       "AAACCCAAGAACTTCC-31  GGACAAGCGCCGGGCCTCAG  GCGGGCCTCAGCGGAACCCA  \n",
       "AAACCCAAGAAGCCAC-34  GGGTGAGGAGCTGGTGGCGT  GCCCAGGGCCGCGAACCCCG  \n",
       "AAACCCAAGAATAGTC-43  GGCCGGCGCCGGATGGAAGG  GGCCGCGCGACGATGGAACG  \n",
       "AAACCCAAGACAGCGT-28  GGGGACGGTTGAGCCTTGGG  GGGTTGAGCCTTGGGAGGGA  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_df = pd.read_csv(crispri_seq)\n",
    "seq_df.set_index('cell_barcode', inplace=True)\n",
    "seq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "464d144a-0183-462e-8dea-529e3da52fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_idx = adata.obs.index.intersection(seq_df.index)\n",
    "adata = adata[common_idx]\n",
    "seq_df = seq_df.loc[common_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f25788ae-fbde-48cd-b7df-8dbbb633d943",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_to_add = ['gene','gene_id','gem_group', 'sgID_AB', 'guide_seq_a', 'guide_seq_b']\n",
    "for c in cols_to_add:\n",
    "    if c in seq_df.columns:\n",
    "        adata.obs[c] = seq_df[c].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4f0afe94-5217-46f7-b3c1-a57515024a4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cov_drug_dose_name</th>\n",
       "      <th>dose_val</th>\n",
       "      <th>control</th>\n",
       "      <th>condition_name</th>\n",
       "      <th>split</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gem_group</th>\n",
       "      <th>sgID_AB</th>\n",
       "      <th>guide_seq_a</th>\n",
       "      <th>guide_seq_b</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGCCAC-34</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>34</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>GGGTGAGGAGCTGGTGGCGT</td>\n",
       "      <td>GCCCAGGGCCGCGAACCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATTCTCTCGAC-42</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>42</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>GGGTGAGGAGCTGGTGGCGT</td>\n",
       "      <td>GCCCAGGGCCGCGAACCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACGGGAGTAATGATG-25</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>25</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>GGGTGAGGAGCTGGTGGCGT</td>\n",
       "      <td>GCCCAGGGCCGCGAACCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGAACAAGCTAGATA-35</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>35</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>GGGTGAGGAGCTGGTGGCGT</td>\n",
       "      <td>GCCCAGGGCCGCGAACCCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGACTCTCTATTGTC-33</th>\n",
       "      <td>UBL5+ctrl</td>\n",
       "      <td>K562</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>1+1</td>\n",
       "      <td>0</td>\n",
       "      <td>K562_UBL5+ctrl_1+1</td>\n",
       "      <td>train</td>\n",
       "      <td>UBL5</td>\n",
       "      <td>ENSG00000198258</td>\n",
       "      <td>33</td>\n",
       "      <td>UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2</td>\n",
       "      <td>GGGTGAGGAGCTGGTGGCGT</td>\n",
       "      <td>GCCCAGGGCCGCGAACCCCG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     condition cell_type  cov_drug_dose_name dose_val control  \\\n",
       "cell_barcode                                                                    \n",
       "AAACCCAAGAAGCCAC-34  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AAAGGATTCTCTCGAC-42  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AACGGGAGTAATGATG-25  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AAGAACAAGCTAGATA-35  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "AAGACTCTCTATTGTC-33  UBL5+ctrl      K562  K562_UBL5+ctrl_1+1      1+1       0   \n",
       "\n",
       "                         condition_name  split  gene          gene_id  \\\n",
       "cell_barcode                                                            \n",
       "AAACCCAAGAAGCCAC-34  K562_UBL5+ctrl_1+1  train  UBL5  ENSG00000198258   \n",
       "AAAGGATTCTCTCGAC-42  K562_UBL5+ctrl_1+1  train  UBL5  ENSG00000198258   \n",
       "AACGGGAGTAATGATG-25  K562_UBL5+ctrl_1+1  train  UBL5  ENSG00000198258   \n",
       "AAGAACAAGCTAGATA-35  K562_UBL5+ctrl_1+1  train  UBL5  ENSG00000198258   \n",
       "AAGACTCTCTATTGTC-33  K562_UBL5+ctrl_1+1  train  UBL5  ENSG00000198258   \n",
       "\n",
       "                     gem_group                                        sgID_AB  \\\n",
       "cell_barcode                                                                    \n",
       "AAACCCAAGAAGCCAC-34         34  UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AAAGGATTCTCTCGAC-42         42  UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AACGGGAGTAATGATG-25         25  UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AAGAACAAGCTAGATA-35         35  UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "AAGACTCTCTATTGTC-33         33  UBL5_-_9938639.23-P1P2|UBL5_+_9938801.23-P1P2   \n",
       "\n",
       "                              guide_seq_a           guide_seq_b  \n",
       "cell_barcode                                                     \n",
       "AAACCCAAGAAGCCAC-34  GGGTGAGGAGCTGGTGGCGT  GCCCAGGGCCGCGAACCCCG  \n",
       "AAAGGATTCTCTCGAC-42  GGGTGAGGAGCTGGTGGCGT  GCCCAGGGCCGCGAACCCCG  \n",
       "AACGGGAGTAATGATG-25  GGGTGAGGAGCTGGTGGCGT  GCCCAGGGCCGCGAACCCCG  \n",
       "AAGAACAAGCTAGATA-35  GGGTGAGGAGCTGGTGGCGT  GCCCAGGGCCGCGAACCCCG  \n",
       "AAGACTCTCTATTGTC-33  GGGTGAGGAGCTGGTGGCGT  GCCCAGGGCCGCGAACCCCG  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63233413-5ce6-48ac-a198-ccf3527b7e10",
   "metadata": {},
   "source": [
    "**Check that all non-control rows have been updated**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b8ba8f67-c352-4553-a18e-f7b289d56aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "is_case = adata.obs['condition'] != 'ctrl'\n",
    "cols_to_check = ['guide_seq_a', 'guide_seq_b']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "22156a4a-8a22-4bcb-8a86-2d8b3bfbd61e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "has_missing_seq = (\n",
    "    adata.obs[cols_to_check].isnull() | \n",
    "    (adata.obs[cols_to_check] == '')\n",
    ").any(axis=1)\n",
    "\n",
    "len(adata.obs[is_case & has_missing_seq])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b5dd4fb-12a6-4fc6-aeb9-68b746178c5c",
   "metadata": {},
   "source": [
    "## Data Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8914679e-7281-4dae-8925-bfd108d4a86b",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_counts = np.array(adata.X.sum(axis=1)).flatten()\n",
    "adata.obs['log_total_counts'] = np.log1p(total_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e2462943-bceb-45ef-ba68-780c1d63f435",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=count_normalize_target)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=n_genes, subset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f77ea077-3f31-4a5c-b744-27798bb0092e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final Feature Space: 5000 Genes\n"
     ]
    }
   ],
   "source": [
    "genes = adata.var.gene_name.tolist()\n",
    "print(f'Final Feature Space: {len(genes)} Genes')\n",
    "\n",
    "with open(data_dir / 'gene_names.json', 'w') as f:\n",
    "    json.dump(genes, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7dc725-39bf-496c-8970-e14d76cb40c5",
   "metadata": {},
   "source": [
    "## Perturbation Handling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0d6261c6-ed48-4a7e-a805-d329c2463215",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_gears_name(name):\n",
    "    # GEARS format is 'Gene+ctrl' -> We want 'Gene'\n",
    "    if name.endswith('+ctrl'):\n",
    "        return name.replace('+ctrl', '')\n",
    "    if name.startswith('ctrl+'):\n",
    "        return name.replace('ctrl+', '')\n",
    "    elif name == 'ctrl':\n",
    "        return 'control'\n",
    "    return str.strip(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ca05552c-3614-45e2-812f-b7f088b9a251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1087, 1087)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_perts = adata.obs['condition'].unique()\n",
    "all_perts = [clean_gears_name(p) for p in all_perts if p != 'ctrl']\n",
    "unique_conditions = sorted(list(set(all_perts)))\n",
    "len(all_perts), len(unique_conditions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d07e2e3-be24-4194-a71a-2bf4978b77f2",
   "metadata": {},
   "source": [
    "**Extract ensembl IDs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9374e52b-a6a3-49c1-882e-7177ff0e5fc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1087"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_ids = adata.obs['gene_id'].unique().tolist()\n",
    "target_ids = [t for t in target_ids if isinstance(t, str) and t.startswith('ENSG')]\n",
    "len(target_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "439089dd-33d9-44e2-ad7c-f6938334ec97",
   "metadata": {},
   "source": [
    "**Query to get uniprot IDs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2dfa0c72-044c-4883-b428-6373164770a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "mg_results = mg.querymany(target_ids, scopes='ensembl.gene,symbol,alias', fields='uniprot,symbol', species='human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e8b7a7cc-faf3-42ae-9cd5-470c2c6b6b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensg_to_acc = {}\n",
    "for res in mg_results:\n",
    "    query_ensg = res['query']\n",
    "    if 'uniprot' in res:\n",
    "        # Prefer Swiss-Prot (Reviewed) -> TrEMBL (Unreviewed)\n",
    "        if 'Swiss-Prot' in res['uniprot']:\n",
    "            val = res['uniprot']['Swiss-Prot']\n",
    "            # Handle list vs string\n",
    "            ensg_to_acc[query_ensg] = val[0] if isinstance(val, list) else val\n",
    "        elif 'TrEMBL' in res['uniprot']:\n",
    "            val = res['uniprot']['TrEMBL']\n",
    "            ensg_to_acc[query_ensg] = val[0] if isinstance(val, list) else val"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bce8c7c-a7eb-4dcb-82b8-c9916d021aa4",
   "metadata": {},
   "source": [
    "**Read uniprot FASTA**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "44c61aa5-7773-438d-987f-2977ea831061",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/home/ubuntu/uniprot/UP000005640_9606.fasta.gz')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_prots = 'https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/UP000005640_9606.fasta.gz'\n",
    "#! wget -P {protein_dir} {all_prots}\n",
    "\n",
    "fasta = [i for i in protein_dir.iterdir()][0]\n",
    "needed_accessions = set(ensg_to_acc.values())\n",
    "acc_to_seq = {}\n",
    "fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d6d7d8d7-6603-4c68-a91f-7cd7b93565b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "with gzip.open(fasta, 'rt') as handle:\n",
    "    for record in SeqIO.parse(handle, 'fasta'):\n",
    "        # Header format: >sp|P12345|GENE_HUMAN ...\n",
    "        # We split by '|' to get P12345 safely\n",
    "        header_parts = record.id.split('|')\n",
    "        \n",
    "        # Check if valid format >db|Accession|...\n",
    "        if len(header_parts) >= 2:\n",
    "            accession = header_parts[1]\n",
    "            \n",
    "            # Only keep if linked to our genes\n",
    "            if accession in needed_accessions:\n",
    "                acc_to_seq[accession] = str(record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "10dc88cd-0073-40df-97dc-a2d4e6723186",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_to_seq = {}\n",
    "missing_targets = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7d56d674-5a41-41e9-93cf-481c0325a709",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ensg in target_ids:\n",
    "    acc = ensg_to_acc.get(ensg)\n",
    "    if acc and acc in acc_to_seq:\n",
    "        id_to_seq[ensg] = acc_to_seq[acc]\n",
    "    else:\n",
    "        missing_targets.append(ensg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a9fab136-2b96-403d-adac-8b95e55e0888",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSG00000196531']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_targets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "272e76b8-6ce2-4a04-b7e9-17c33f05ed1c",
   "metadata": {},
   "source": [
    "**Search NCBI for missing genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "92fd176b-b936-44ea-9d53-5fac89a536a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = 'gptomics@gmail.com'\n",
    "results = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "190ae4f4-7936-4f6b-84fe-0a85635ff802",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.05it/s]\n"
     ]
    }
   ],
   "source": [
    "for ensg_id in tqdm(missing_targets.copy()):\n",
    "    try:\n",
    "        search_handle = Entrez.esearch(db=\"gene\", term=ensg_id, retmax=1)\n",
    "        search_record = Entrez.read(search_handle)\n",
    "        \n",
    "        if search_record['IdList']:\n",
    "            ncbi_gene_id = search_record['IdList'][0]\n",
    "            \n",
    "            # Step 2: Link Gene -> Protein (RefSeq)\n",
    "            # This finds the \"representative\" protein for this gene\n",
    "            link_handle = Entrez.elink(dbfrom=\"gene\", db=\"protein\", id=ncbi_gene_id, linkname=\"gene_protein_refseq\")\n",
    "            link_record = Entrez.read(link_handle)\n",
    "            \n",
    "            # Check if links exist\n",
    "            if link_record and link_record[0]['LinkSetDb']:\n",
    "                # Get the first linked Protein ID\n",
    "                protein_id = link_record[0]['LinkSetDb'][0]['Link'][0]['Id']\n",
    "                \n",
    "                # Step 3: Fetch Sequence\n",
    "                fetch_handle = Entrez.efetch(db=\"protein\", id=protein_id, rettype=\"fasta\", retmode=\"text\")\n",
    "                seq_record = SeqIO.read(fetch_handle, \"fasta\")\n",
    "                \n",
    "                results[ensg_id] = str(seq_record.seq)\n",
    "                missing_targets.remove(ensg_id)\n",
    "            else:\n",
    "                results[ensg_id] = 'M' # Found gene, but no RefSeq protein linked\n",
    "        else:\n",
    "            results[ensg_id] = 'M' \n",
    "    except Exception as e:\n",
    "        print(f'Failed Entrez lookup for {ensg_id}: {e}')\n",
    "        results[ensg_id] = 'M'\n",
    "id_to_seq.update(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "290348c5-e3e8-482a-a2a5-6db276aa92a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Initial: 1087 | Found 1087 | Missing 0'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "found = len([v for v in id_to_seq.values() if v != 'M'])\n",
    "\n",
    "f'Initial: {len(target_ids)} | Found {found} | Missing {len(missing_targets)}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77640c76-00db-4999-a8e1-4e13fc092c8d",
   "metadata": {},
   "source": [
    "**Saving Perturbations**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6f83160c-de36-4aac-b407-00b0d5330cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pert_dir /'perturbation_seq.json', 'w') as f:\n",
    "    json.dump({str(k): str(v) for k, v in id_to_seq.items()}, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bffe6d5-e5dd-4cdf-a485-ee0a9ded7b6b",
   "metadata": {},
   "source": [
    "### Get Embeddings for Perturbations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bda5602a-8d3d-4d57-ba5e-61df9214b274",
   "metadata": {},
   "source": [
    "**Get Protein Embeddings**\n",
    "for protein embeddings, we'll use the [ESM-2](https://huggingface.co/facebook/esm2_t6_8M_UR50D) for now given it's been a common workhorse. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3e647c60-1427-4c9e-a769-54be5db97a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_model_name = 'facebook/esm2_t6_8M_UR50D' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4c2b3a3f-466b-47df-a038-16c681914cc6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f5967a9419414d86b52941f101319bdf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer_config.json:   0%|          | 0.00/95.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bcc5de8e52474ef9951206e35280422d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "vocab.txt:   0%|          | 0.00/93.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d038300265a24f4298afa9e6285610a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c4b103bafeb04e80a7a6cf926de9285e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "config.json:   0%|          | 0.00/775 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8cb81f4d06dc4de4a62f0e67bc3a527a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "model.safetensors:   0%|          | 0.00/31.4M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of EsmModel were not initialized from the model checkpoint at facebook/esm2_t6_8M_UR50D and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "EsmModel(\n",
       "  (embeddings): EsmEmbeddings(\n",
       "    (word_embeddings): Embedding(33, 320, padding_idx=1)\n",
       "    (dropout): Dropout(p=0.0, inplace=False)\n",
       "  )\n",
       "  (encoder): EsmEncoder(\n",
       "    (layer): ModuleList(\n",
       "      (0-5): 6 x EsmLayer(\n",
       "        (attention): EsmAttention(\n",
       "          (self): EsmSelfAttention(\n",
       "            (query): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (key): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (value): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (rotary_embeddings): RotaryEmbedding()\n",
       "          )\n",
       "          (output): EsmSelfOutput(\n",
       "            (dense): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (dropout): Dropout(p=0.0, inplace=False)\n",
       "          )\n",
       "          (LayerNorm): LayerNorm((320,), eps=1e-05, elementwise_affine=True)\n",
       "        )\n",
       "        (intermediate): EsmIntermediate(\n",
       "          (dense): Linear(in_features=320, out_features=1280, bias=True)\n",
       "        )\n",
       "        (output): EsmOutput(\n",
       "          (dense): Linear(in_features=1280, out_features=320, bias=True)\n",
       "          (dropout): Dropout(p=0.0, inplace=False)\n",
       "        )\n",
       "        (LayerNorm): LayerNorm((320,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "    )\n",
       "    (emb_layer_norm_after): LayerNorm((320,), eps=1e-05, elementwise_affine=True)\n",
       "  )\n",
       "  (pooler): EsmPooler(\n",
       "    (dense): Linear(in_features=320, out_features=320, bias=True)\n",
       "    (activation): Tanh()\n",
       "  )\n",
       "  (contact_head): EsmContactPredictionHead(\n",
       "    (regression): Linear(in_features=120, out_features=1, bias=True)\n",
       "    (activation): Sigmoid()\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_tokenizer = AutoTokenizer.from_pretrained(prot_model_name)\n",
    "prot_model = AutoModel.from_pretrained(prot_model_name)\n",
    "prot_model.to(device).eval()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "320dd1ff-2b66-40dd-87f1-ac1aa09681c4",
   "metadata": {},
   "source": [
    "**Get Embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "20bb1f09-0322-4b05-a1a3-fabe037bbbcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_vectors = []\n",
    "prot_index_map = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "2c3511c5-fce8-4bcf-8376-19127a615f83",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████| 1087/1087 [00:08<00:00, 126.62it/s]\n"
     ]
    }
   ],
   "source": [
    "with torch.no_grad():\n",
    "    for idx, (ensg, seq) in enumerate(tqdm(id_to_seq.items())):\n",
    "        inputs = prot_tokenizer(seq, return_tensors='pt', truncation=True, max_length=1024).to(device)\n",
    "        out = prot_model(**inputs)\n",
    "        emb = out.last_hidden_state[0].mean(dim=0).cpu().numpy()\n",
    "        \n",
    "        prot_vectors.append(emb)\n",
    "        prot_index_map[ensg] = idx # Mapping is now ENSG -> Index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49bc3dbd-4183-4ea2-812e-0d47996788ab",
   "metadata": {},
   "source": [
    "**Save Protein Mapping/Embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "208e083d-0595-4a20-93de-f56b7c702928",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_bank = np.vstack(prot_vectors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5075aa69-ecb0-41f0-ac2d-8aa127c60567",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(pert_dir / 'action_embeddings_esm2.npy', final_bank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0017ab82-40ed-4fa0-abf7-17b9469a8a11",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pert_dir / 'anchor_to_id.json', 'w') as f:\n",
    "    json.dump(prot_index_map, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccf190de-b944-4ee7-90e6-794234aa7c88",
   "metadata": {},
   "source": [
    "**Get sgRNA embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "7ec22e6c-f5ae-4e35-a7ea-89233198f811",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_model_name = 'InstaDeepAI/NTv3_650M_pre' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2e34f26b-7b71-4283-b95c-b0bbf95a0f87",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ce76d2f0baf1489cbe8d2f47f3f9fe4e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer_config.json:   0%|          | 0.00/1.48k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4c17445c01ea4720abf6cf44fa7db058",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenization_ntv3.py:   0%|          | 0.00/7.85k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "A new version of the following files was downloaded from https://huggingface.co/InstaDeepAI/ntv3_base_model:\n",
      "- tokenization_ntv3.py\n",
      ". Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a6018faa51d24113b3f55e9dbfc4f56b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "vocab.json:   0%|          | 0.00/138 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "caf3de7052a146659236b82e38206ef8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "special_tokens_map.json:   0%|          | 0.00/149 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fd2fef161d3947748ca4303c8d9e3219",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "config.json:   0%|          | 0.00/1.75k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d03264dfc0f340fab62038fa99091b1b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "configuration_ntv3_pretrained.py:   0%|          | 0.00/8.09k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "A new version of the following files was downloaded from https://huggingface.co/InstaDeepAI/ntv3_base_model:\n",
      "- configuration_ntv3_pretrained.py\n",
      ". Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "06e8647ea00744faae2652d9a9f95c46",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "modeling_ntv3_pretrained.py:   0%|          | 0.00/35.2k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "A new version of the following files was downloaded from https://huggingface.co/InstaDeepAI/ntv3_base_model:\n",
      "- modeling_ntv3_pretrained.py\n",
      ". Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a0e039a1b620465480904a0ce59540ab",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "model.safetensors:   0%|          | 0.00/2.61G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "NTv3PreTrained(\n",
       "  (core): Core(\n",
       "    (embed_layer): Embedding(11, 16)\n",
       "    (stem): Stem(\n",
       "      (conv): Conv1d(16, 1536, kernel_size=(15,), stride=(1,), padding=same)\n",
       "    )\n",
       "    (conv_tower_blocks): ModuleList(\n",
       "      (0-6): 7 x ConvTowerBlock(\n",
       "        (conv): ConvBlock(\n",
       "          (conv): Conv1d(1536, 1536, kernel_size=(5,), stride=(1,), padding=same)\n",
       "          (layer_norm): LayerNormFP32((np.int64(1536),), eps=1e-05, elementwise_affine=True)\n",
       "        )\n",
       "        (res_conv): ResidualConvBlock(\n",
       "          (conv_block): ConvBlock(\n",
       "            (conv): Conv1d(1536, 1536, kernel_size=(1,), stride=(1,), padding=same)\n",
       "            (layer_norm): LayerNormFP32((np.int64(1536),), eps=1e-05, elementwise_affine=True)\n",
       "          )\n",
       "        )\n",
       "        (avg_pool): AvgPool1d(kernel_size=(2,), stride=(2,), padding=(0,))\n",
       "      )\n",
       "    )\n",
       "    (transformer_blocks): ModuleList(\n",
       "      (0-11): 12 x SelfAttentionBlock(\n",
       "        (self_attention_layer_norm): LayerNormFP32((1536,), eps=1e-05, elementwise_affine=True)\n",
       "        (final_layer_norm): LayerNormFP32((1536,), eps=1e-05, elementwise_affine=True)\n",
       "        (sa_layer): MultiHeadAttention(\n",
       "          (query_head): LinearProjectionHeInit(\n",
       "            (linear): Linear(in_features=1536, out_features=1536, bias=True)\n",
       "          )\n",
       "          (key_head): LinearProjectionHeInit(\n",
       "            (linear): Linear(in_features=1536, out_features=1536, bias=True)\n",
       "          )\n",
       "          (value_head): LinearProjectionHeInit(\n",
       "            (linear): Linear(in_features=1536, out_features=1536, bias=True)\n",
       "          )\n",
       "          (mha_output): Linear(in_features=1536, out_features=1536, bias=True)\n",
       "          (rotary_embedding): RotaryEmbedding()\n",
       "        )\n",
       "        (fc1): Linear(in_features=1536, out_features=12288, bias=False)\n",
       "        (fc2): Linear(in_features=6144, out_features=1536, bias=False)\n",
       "        (_ffn_activation_fn): SiLU()\n",
       "      )\n",
       "    )\n",
       "    (deconv_tower_blocks): ModuleList(\n",
       "      (0-6): 7 x DeconvTowerBlock(\n",
       "        (conv): UpsamplingDeconvBlock(\n",
       "          (conv): Conv1d(1536, 1536, kernel_size=(5,), stride=(1,), padding=same)\n",
       "          (layer_norm): LayerNormFP32((np.int64(1536),), eps=1e-05, elementwise_affine=True)\n",
       "        )\n",
       "        (res_conv): ResidualUpsamplingDeconvBlock(\n",
       "          (conv_block): UpsamplingDeconvBlock(\n",
       "            (conv): ConvTranspose1d(1536, 1536, kernel_size=(1,), stride=(1,))\n",
       "            (layer_norm): LayerNormFP32((np.int64(1536),), eps=1e-05, elementwise_affine=True)\n",
       "          )\n",
       "        )\n",
       "      )\n",
       "    )\n",
       "    (lm_head): ModuleDict(\n",
       "      (hidden_layers): ModuleList()\n",
       "      (head): Linear(in_features=1536, out_features=11, bias=True)\n",
       "    )\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_tokenizer = AutoTokenizer.from_pretrained(dna_model_name, trust_remote_code=True)\n",
    "dna_model = AutoModelForMaskedLM.from_pretrained(dna_model_name, trust_remote_code=True)\n",
    "dna_model.to(device).eval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "32f03d0c-46ab-4c9c-ab10-6fa2464b9ca2",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_pairs = adata.obs[['sgID_AB', 'guide_seq_a', 'guide_seq_b']].drop_duplicates().dropna()\n",
    "dna_vectors = []\n",
    "dna_index_map = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3d19f274-0436-45c2-beaa-741abc7ff203",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ntv3_embedding(sequence, tokenizer, model, device):\n",
    "    # Tokenize with PADDING to 128\n",
    "    inputs = tokenizer(\n",
    "        [sequence], \n",
    "        return_tensors=\"pt\",\n",
    "        add_special_tokens=False, \n",
    "        padding=\"max_length\",    \n",
    "        max_length=128,          \n",
    "    )\n",
    "    \n",
    "    # Manual Mask Generation (Fixes KeyError)\n",
    "    pad_id = tokenizer.pad_token_id if tokenizer.pad_token_id is not None else 0\n",
    "    attention_mask = (inputs['input_ids'] != pad_id).long()\n",
    "    inputs = {k: v.to(device) for k, v in inputs.items() if k != 'attention_mask'} \n",
    "    attention_mask = attention_mask.to(device)\n",
    "    \n",
    "    # Forward Pass\n",
    "    with torch.no_grad():\n",
    "        outputs = model(**inputs, output_hidden_states=True)\n",
    "        \n",
    "    # Masked Mean Pooling\n",
    "    token_embeddings = outputs.hidden_states[-1] # Shape: [1, 128, 1536]\n",
    "    \n",
    "    # Expand mask to match embedding dimensions \n",
    "    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()\n",
    "    \n",
    "    # Sum valid tokens / Count valid tokens\n",
    "    sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)\n",
    "    \n",
    "    # Avoid division by zero\n",
    "    sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)\n",
    "    \n",
    "    # Result is [1, 1536] -> squeeze to [1536]\n",
    "    return (sum_embeddings / sum_mask).squeeze(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "26e4181e-98ce-403f-af54-6929c5113154",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████| 1250/1250 [00:53<00:00, 23.17it/s]\n"
     ]
    }
   ],
   "source": [
    "with torch.no_grad():\n",
    "    for i, (barcode, row) in enumerate(tqdm(guide_pairs.iterrows(), total=len(guide_pairs))):\n",
    "        sg_id = row['sgID_AB']\n",
    "        seq_a = row['guide_seq_a'].upper()\n",
    "        seq_b = row['guide_seq_b'].upper()\n",
    "        \n",
    "        # Embed A & B using the helper\n",
    "        emb_a = get_ntv3_embedding(seq_a, dna_tokenizer, dna_model, device)\n",
    "        emb_b = get_ntv3_embedding(seq_b, dna_tokenizer, dna_model, device)\n",
    "        \n",
    "        # Average (still on GPU)\n",
    "        combined_tensor = (emb_a + emb_b) / 2\n",
    "        \n",
    "        # Move to CPU/Numpy for storage\n",
    "        combined = combined_tensor.cpu().numpy()\n",
    "        \n",
    "        dna_vectors.append(combined)\n",
    "        dna_index_map[sg_id] = i"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46523039-1da8-4fac-8994-b14bfa865235",
   "metadata": {},
   "source": [
    "**Save sgRNA embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "725325d3-c090-465f-bfa0-79925520f98a",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(pert_dir / 'input_embeddings_dna.npy', np.vstack(dna_vectors))\n",
    "with open(pert_dir / 'input_to_id.json', 'w') as f:\n",
    "    json.dump(dna_index_map, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4baca5c5-72d4-40ba-91ba-aafc4e74ad91",
   "metadata": {},
   "source": [
    "## Create Pertrubation Dataset\n",
    "\n",
    "we end up with a longer list than targets since we have different cripri targets to the same genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "55cb8424-8198-40b0-a3f6-e24ad586019f",
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_pairs = []\n",
    "pert_metadata = {} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7d2d62f6-c040-410e-91a8-7aed63fa5b8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████| 1250/1250 [00:05<00:00, 219.97it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1153, 1087)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for i, (barcode, row) in enumerate(tqdm(guide_pairs.iterrows(), total=len(guide_pairs))):\n",
    "    sg_id = row['sgID_AB']\n",
    "    matching = adata.obs[adata.obs['sgID_AB'] == sg_id]\n",
    "    if len(matching) == 0: continue\n",
    "    \n",
    "    target_ensg = matching['gene_id'].iloc[0]\n",
    "    \n",
    "    if target_ensg in prot_index_map and sg_id in dna_index_map:\n",
    "        this_mod = modality_to_id['dna']\n",
    "        this_mode = mode_to_id['crispri']\n",
    "        \n",
    "        # 1. Add to Alignment Training Pairs\n",
    "        valid_pairs.append((\n",
    "            dna_index_map[sg_id], \n",
    "            prot_index_map[target_ensg], \n",
    "            this_mod, \n",
    "            this_mode\n",
    "        ))\n",
    "        \n",
    "        # 2. Register for Sharding Step\n",
    "        pert_metadata[sg_id] = (this_mod, this_mode)\n",
    "        \n",
    "len(valid_pairs), len(target_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "8d5cb8fe-2926-4ee3-be28-f1bcb7f2c243",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_alignment(pairs, name):\n",
    "    if not pairs: return\n",
    "    inp, anch, mod, mode = zip(*pairs)\n",
    "    np.savez(pert_dir / name / f'pert_pairs_cripri_{name}.npz', \n",
    "             input_idx=np.array(inp, dtype=np.int32), \n",
    "             anchor_idx=np.array(anch, dtype=np.int32),\n",
    "             modality=np.array(mod, dtype=np.int8), \n",
    "             mode=np.array(mode, dtype=np.int8))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8c8c295c-a22a-4954-9120-718ba4ac5f1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_pairs, val_pairs = train_test_split(valid_pairs, test_size=0.1, random_state=42)\n",
    "save_alignment(train_pairs, 'train')\n",
    "save_alignment(val_pairs, 'val')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be255d96-9645-496a-8f55-68225562445f",
   "metadata": {},
   "source": [
    "## Prepare Controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c98982a9-79d1-4e87-b33e-0ed65b11549b",
   "metadata": {},
   "outputs": [],
   "source": [
    "control_groups = {}\n",
    "ctrl_inds = np.where(adata.obs['condition'] == 'ctrl')[0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b86f5eef-8b0a-4471-ac50-2e4f97803237",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx in ctrl_inds:\n",
    "    grp = adata.obs.iloc[idx]['gem_group']\n",
    "    if grp not in control_groups: control_groups[grp] = {'X': [], 'total': []}\n",
    "    control_groups[grp]['X'].append(adata.X[idx].toarray().flatten())\n",
    "    control_groups[grp]['total'].append(adata.obs['log_total_counts'].iloc[idx])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "880f10c5-209e-4889-8ead-9ae6894c0d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in control_groups:\n",
    "    control_groups[g]['X'] = np.array(control_groups[g]['X'], dtype=np.float32)\n",
    "    control_groups[g]['total'] = np.array(control_groups[g]['total'], dtype=np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "0d5a2762-35dd-4921-a265-39073ad1dbed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 48 control cells.\n"
     ]
    }
   ],
   "source": [
    "print(f'Found {len(control_groups.keys())} control cells.')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc6e4054-d83e-43c7-86b9-974cd1358fe6",
   "metadata": {},
   "source": [
    "## Shard Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e346c31c-bc57-4467-94be-cc56022e6a4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_shards(split_name, conditions, ds_name):\n",
    "    \"\"\"\n",
    "    Iterates through cells belonging to the given conditions, \n",
    "    pairs them with random controls, and saves .npz shards.\n",
    "    \"\"\"\n",
    "    print(f'Split: {split_name.upper()}')\n",
    "    \n",
    "    # Filter cells belonging to these perturbations\n",
    "    mask = adata.obs['condition'].isin(conditions) & (adata.obs['condition'] != 'ctrl')\n",
    "    indices = np.where(mask)[0]\n",
    "    np.random.shuffle(indices)\n",
    "    \n",
    "    # Buffer for current shard\n",
    "    buffer = {k: [] for k in ['control_x', 'control_total', 'case_x', 'case_total', 'pert_idx', 'pert_mod', 'pert_mode']}\n",
    "    \n",
    "    shard_count = 0\n",
    "    save_path = tok_dir / split_name\n",
    "    \n",
    "    for idx in tqdm(indices):\n",
    "        # get group/pert info\n",
    "        grp = adata.obs.iloc[idx]['gem_group']\n",
    "        sg_id = adata.obs.iloc[idx]['sgID_AB']\n",
    "\n",
    "        # validate exists in controls and perturbations\n",
    "        if grp not in control_groups or sg_id not in pert_metadata: \n",
    "            print(f'skipping {idx} |grp {grp} |sg_id {sg_id}')\n",
    "            continue\n",
    "        \n",
    "        # extract perturbation metadata\n",
    "        p_mod, p_mode = pert_metadata[sg_id]\n",
    "\n",
    "        # Sample Control\n",
    "        c_bank = control_groups[grp]\n",
    "        ri = np.random.randint(len(c_bank['X']))\n",
    "        \n",
    "        # 3. Add to Buffer\n",
    "        buffer['control_x'].append(c_bank['X'][ri])\n",
    "        buffer['control_total'].append(c_bank['total'][ri])\n",
    "        buffer['case_x'].append(adata.X[idx].toarray().flatten())\n",
    "        buffer['case_total'].append(adata.obs['log_total_counts'].iloc[idx])\n",
    "        buffer['pert_idx'].append(dna_index_map[sg_id])\n",
    "        buffer['pert_mod'].append(p_mod)  \n",
    "        buffer['pert_mode'].append(p_mode)\n",
    "        \n",
    "        # 4. Save if buffer full\n",
    "        if len(buffer['case_x']) >= chunk_size:\n",
    "            np.savez(save_path/ f'shard_{ds_name}_{split_name}_{shard_count:04d}.npz',\n",
    "                     control=np.array(buffer['control_x'], dtype=np.float32),\n",
    "                     control_total=np.array(buffer['control_total'], dtype=np.float32),\n",
    "                     case=np.array(buffer['case_x'], dtype=np.float32),\n",
    "                     case_total=np.array(buffer['case_total'], dtype=np.float32),\n",
    "                     pert_idx=np.array(buffer['pert_idx'], dtype=np.int32),\n",
    "                     pert_modality=np.array(buffer['pert_mod'], dtype=np.int8),\n",
    "                     pert_mode=np.array(buffer['pert_mode'], dtype=np.int8))\n",
    "            buffer = {k: [] for k in buffer}\n",
    "            shard_count += 1\n",
    "            \n",
    "    # Save leftovers\n",
    "    if buffer['case_x']:\n",
    "         np.savez(save_path / f'shard_{ds_name}_{split_name}_{shard_count:04d}.npz',\n",
    "                     control=np.array(buffer['control_x'], dtype=np.float32),\n",
    "                     control_total=np.array(buffer['control_total'], dtype=np.float32),\n",
    "                     case=np.array(buffer['case_x'], dtype=np.float32),\n",
    "                     case_total=np.array(buffer['case_total'], dtype=np.float32),\n",
    "                     pert_idx=np.array(buffer['pert_idx'], dtype=np.int32),\n",
    "                     pert_modality=np.array(buffer['pert_mod'], dtype=np.int8),\n",
    "                     pert_mode=np.array(buffer['pert_mode'], dtype=np.int8))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "482f11fc-ae7e-4c84-846c-eab28f6ceb0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_map = pert_data.set2conditions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "6a0bf447-aecf-4c4b-8ef1-6cbe398156b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Split: TRAIN\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████| 101682/101682 [00:36<00:00, 2814.64it/s]\n"
     ]
    }
   ],
   "source": [
    "write_shards('train', split_map['train'], dataset_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "6f659adb-efce-44e0-a628-2c4eb6f75d05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Split: VAL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████| 11044/11044 [00:03<00:00, 3009.56it/s]\n"
     ]
    }
   ],
   "source": [
    "write_shards('val', split_map['val'], dataset_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "21054db1-434c-4c50-a38d-b6094ba5a952",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Split: TEST\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████| 38829/38829 [00:13<00:00, 2900.80it/s]\n"
     ]
    }
   ],
   "source": [
    "write_shards('test', split_map['test'], dataset_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6d94654-0ba8-4c11-aba3-a9588069f255",
   "metadata": {},
   "source": [
    "## Write Pretraining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f084af46-dab0-45a7-b9eb-83fb1cd6be94",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_chunk_size = 50000     \n",
    "pt_tok_dir = data_dir / 'pretraining'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "80ff306e-70fa-4847-bf4c-e0cd90b57d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_pt_shards(split_name, condition_list, ds_name):\n",
    "    \"\"\"\n",
    "    Iterates through cells belonging to the given conditions, \n",
    "    pairs them with random controls, and saves .npz shards.\n",
    "    \"\"\"\n",
    "    print(f'Split: {split_name.upper()}')\n",
    "    \n",
    "    # Filter cells belonging to these split\n",
    "    mask = adata.obs['condition'].isin(condition_list)\n",
    "    indices = np.where(mask)[0]\n",
    "    \n",
    "    # Shuffle for randomness\n",
    "    np.random.shuffle(indices)\n",
    "    \n",
    "    # Buffer for current shard\n",
    "    buffer = {\n",
    "        'x': [], \n",
    "        'total': []\n",
    "    }\n",
    "    \n",
    "    shard_count = 0\n",
    "    save_path = pt_tok_dir / split_name\n",
    "    \n",
    "    for idx in tqdm(indices):\n",
    "        # 1. Get Data\n",
    "        x = adata.X[idx].toarray().flatten().astype(np.float32)\n",
    "        total = adata.obs['log_total_counts'].iloc[idx].astype(np.float32)\n",
    "        \n",
    "        # 3. Add to Buffer\n",
    "        buffer['x'].append(x)\n",
    "        buffer['total'].append(total)\n",
    "        \n",
    "        # 4. Save if buffer full\n",
    "        if len(buffer['x']) >= pt_chunk_size:\n",
    "            np.savez(\n",
    "                save_path / f'pt_shard_{ds_name}_{split_name}_{shard_count:04d}.npz',\n",
    "                x=np.array(buffer['x']),\n",
    "                total=np.array(buffer['total'])\n",
    "            )\n",
    "            # Reset\n",
    "            buffer = {k: [] for k in buffer}\n",
    "            shard_count += 1\n",
    "            \n",
    "    # Save leftovers\n",
    "    if len(buffer['x']) > 0:\n",
    "        np.savez(\n",
    "                save_path / f'pt_shard_{ds_name}_{split_name}_{shard_count:04d}.npz',\n",
    "                x=np.array(buffer['x']),\n",
    "                total=np.array(buffer['total'])\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "2eea29ea-9ad3-4563-a365-8f086c70a53a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Split: TRAIN\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████| 112373/112373 [00:04<00:00, 23088.77it/s]\n"
     ]
    }
   ],
   "source": [
    "write_pt_shards('train', split_map['train'], dataset_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f9955a45-ea16-42d3-9d8f-8c71b3c635f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Split: VAL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████| 11044/11044 [00:00<00:00, 30547.30it/s]\n"
     ]
    }
   ],
   "source": [
    "write_pt_shards('val', split_map['val'], dataset_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7329b7fb-826e-41e8-a37d-a7f4205de437",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
