{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8e04f6e-559d-49c8-967c-60576dcdf2b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from gears import PertData, GEARS\n",
    "import scanpy as sc\n",
    "import json\n",
    "import gzip\n",
    "import mygene\n",
    "from Bio import Entrez, SeqIO\n",
    "from tqdm import tqdm\n",
    "from transformers import AutoTokenizer, AutoModel\n",
    "import torch\n",
    "import numpy as np\n",
    "\n",
    "#import pandas as pd\n",
    "# import gseapy as gp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1e342faa-37db-4936-8e1e-e3508894041b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path('/Users/djemec/data/jepa/v0_3')\n",
    "raw_dir = data_dir / 'raw_files'\n",
    "pert_dir = data_dir / 'pert_embd'\n",
    "protein_dir = data_dir / 'uniprot'\n",
    "\n",
    "\n",
    "#protein_dir = data_dir / 'uniprot'\n",
    "#train_dir = data_dir / 'training'\n",
    "#pt_dir = data_dir / 'pretraining'\n",
    "splits= ['train','val','test']\n",
    "\n",
    "\n",
    "n_genes = 8192 # 2**13\n",
    "count_normalize_target = 1e4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8a723ed-c50a-423d-98c5-4522cf0a6a52",
   "metadata": {},
   "source": [
    "**Breakdown of perturbation**\n",
    "\n",
    "* Replogle K562 Ess - CRISPRi (Repression)\n",
    "* Replogle K562 GW - CRISPRi (Repression)\n",
    "* Replogle RPE1 - CRISPRi (Repression)\n",
    "* Adamson - CRISPRi (Repression)\n",
    "* Norman - CRISPRa (Activation) + Combinatorial\n",
    "* SciPlex 3 - Drug Treatment\n",
    "\n",
    "Because of this breakdown, for now we'll just use the CRISPRi datasets.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3d56d9d8-8484-4402-8ae6-ca18c9ddc1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets= {\n",
    "    'k562e':raw_dir / 'k562e',\n",
    "    'k562gw':raw_dir /'k562gw'/'replogle_k562_gw_expanded_8k.h5ad',\n",
    "    'rep1e':raw_dir /'rep1e'/'perturb_processed.h5ad',\n",
    "    'adamson':raw_dir /'adamson'/'perturb_processed.h5ad'\n",
    "}\n",
    "total_cells = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c9ec5af-18bd-4179-b08e-c452a8a12f68",
   "metadata": {},
   "source": [
    "## Extract Anchor File - K562 essential\n",
    "we'll build out 4 important variables from our key dataset.  k562 will be the GEARS dataset we test against and so we need to be extra careful with data leakage on this set. \n",
    "1. `all_genes` - all genes with expression we'll have in our model\n",
    "2. `anchor_genes` - minimum set of genes based on k562-essential. We have to have these to be able to do validation\n",
    "3. `all_perts`- all perturbations we've seen that we'll train on minus the forbidden perturbations\n",
    "4. `forbidden_perturbations` - perturbations we cannot include from the k562 dataset for our validation. If we do, we'll \"leak\" for our validation\n",
    "\n",
    "We'll also track how many cells we're training on in general.  note that this will include all the cells in test/train/val so we'll actually only train on a subset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0025dc1-e77e-42fe-84ef-65323a690f6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n",
      "Found local copy...\n",
      "Found local copy...\n",
      "These perturbations are not in the GO graph and their perturbation can thus not be predicted\n",
      "['C7orf26+ctrl' 'C14orf178+ctrl' 'RPS10-NUDT3+ctrl' 'SEM1+ctrl' 'FAU+ctrl']\n",
      "Local copy of pyg dataset is detected. Loading...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "pert_data = PertData(datasets['k562e']) \n",
    "pert_data.load(data_name='replogle_k562_essential')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d944e14-83f1-4523-8bf2-f39e2a925870",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Local copy of split is detected. Loading...\n",
      "Simulation split test composition:\n",
      "combo_seen0:0\n",
      "combo_seen1:0\n",
      "combo_seen2:0\n",
      "unseen_single:272\n",
      "Done!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "here1\n"
     ]
    }
   ],
   "source": [
    "pert_data.prepare_split(split='simulation', seed=1) \n",
    "adata = pert_data.adata "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "db749d42-b84c-4eb7-9b07-dd3008802351",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_gears_name(name):\n",
    "    # GEARS format is 'Gene+ctrl' -> We want 'Gene'\n",
    "    if name.endswith('+ctrl'):\n",
    "        return name.replace('+ctrl', '')\n",
    "    if name.startswith('ctrl+'):\n",
    "        return name.replace('ctrl+', '')\n",
    "    elif name == 'ctrl':\n",
    "        return 'control'\n",
    "    return str.strip(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4f529443-d41c-447e-af1b-ce11cd89a2cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "272"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hold_out_perts = set(pert_data.set2conditions['test'])\n",
    "hold_out_perts = set([clean_gears_name(i) for i in hold_out_perts])\n",
    "len(hold_out_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9fff3cdf-ff55-41a1-88e6-e9bb9d80b81d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=count_normalize_target)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=n_genes, subset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "75b2c85a-775c-4174-bafd-a77c2f230927",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5000, 4999)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anchor_genes = adata.var.gene_name.tolist()\n",
    "all_genes = sorted(list(set(anchor_genes)))\n",
    "len(anchor_genes), len(all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b05076dd-4165-4ce0-ab3b-e9a24b7fc608",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1087"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_perts = adata.obs['condition'].unique()\n",
    "all_perts = [clean_gears_name(p) for p in all_perts if p != 'ctrl']\n",
    "all_perts = sorted(list(set(all_perts)))\n",
    "len(all_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8648d859-ea64-4075-b3ba-d247e51fe01d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(162246, 162246)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k562e_total_cells = adata.n_obs\n",
    "total_cells += k562e_total_cells\n",
    "k562e_total_cells, total_cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe835716-7b27-4c99-834d-a55246da71a5",
   "metadata": {},
   "source": [
    "## Add K562 genome wide\n",
    "we'd previously already filtered down the K562 genome wide dataset based on forbidden perturbations and our target gene count so for this dataset it should be a matter of reading the file and updating our variables.  we'll be adding to:\n",
    "1. `all_genes` until we reach our limit\n",
    "2. `all_perts` since this is a k562 dataset we'll make sure to not add any forbidden perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f1b0ad79-1327-436d-9540-dcaaa06e6b54",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_gw = sc.read_h5ad(datasets['k562gw'], backed='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1742be01-c733-478d-af8f-964238881ded",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1943010, 2105256)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k652gw_total_cells = adata_gw.n_obs\n",
    "total_cells += k652gw_total_cells\n",
    "k652gw_total_cells, total_cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20692bb0-b430-413b-9295-02f3c08e1446",
   "metadata": {},
   "source": [
    "**Unify Genes** Now we'll merge the genes we extracted with the anchor_genes.  Since we preprocessed the genome-wide dataset we know that this will come up to exactly our target gene count.  in the future we'd have to be careful to only add up to the target with each dataset to not explode our model size. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a4ba0dfb-11d5-4677-9088-1c1f75f62771",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8192,\n",
       " ['A1BG',\n",
       "  'AAAS',\n",
       "  'AACS',\n",
       "  'AAGAB',\n",
       "  'AAK1',\n",
       "  'AAMDC',\n",
       "  'AAMP',\n",
       "  'AAR2',\n",
       "  'AARS',\n",
       "  'AARS2'])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gw_genes = adata_gw.var_names.tolist()\n",
    "gw_genes = sorted(list(set(gw_genes)))\n",
    "len(gw_genes), gw_genes[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0f49f2e0-a503-4a34-8aff-4151c89b5882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8192, True)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genes = set(all_genes).union(set(gw_genes))\n",
    "len(all_genes), len(all_genes) <= n_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d330c9-6cac-49ee-ad78-e33697a766b6",
   "metadata": {},
   "source": [
    "**Unify Perturbations** \n",
    "\n",
    "Now we'll merge the perturbations we extracted with the previous perturbations.  We need to make sure to flag perturbations we don't want "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "64c17608-ca49-4e0b-8dfa-3d21f7ee9d60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9594"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gw_perts = adata_gw.obs['perturbation'].unique()\n",
    "gw_perts = [clean_gears_name(p) for p in gw_perts if p != 'control']\n",
    "gw_perts = sorted(list(set(gw_perts)))\n",
    "len(gw_perts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b702f3b-7a8e-43fc-aaef-3103c565e3c1",
   "metadata": {},
   "source": [
    "check that no forbidden perturbations leaked before merging the perturbations in. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c1e6c8f8-2794-4a85-948c-9db5a8d1f126",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(hold_out_perts) & set(gw_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3bdeab1f-c07c-4ec4-9b20-49d11affff32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9869"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_perts = set(all_perts).union(set(gw_perts))\n",
    "len(all_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e67dd440-a850-4db7-84a7-987a7277103f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'total cells: 2105256 | total genes 8192 | total perturbations 9869'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'total cells: {total_cells} | total genes {len(all_genes)} | total perturbations {len(all_perts)}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "366f8f13-82b4-4032-ac46-b89176b52a93",
   "metadata": {},
   "source": [
    "## rep1e Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4210ecc0-d239-42d4-9607-7c93b861a92f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_adata = sc.read_h5ad(datasets['rep1e'], backed='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c419fab4-c019-4bdd-bc6a-691771f85996",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(162733, 2267989)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_total_cells = ds_adata.n_obs\n",
    "total_cells += ds_total_cells\n",
    "ds_total_cells, total_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "05e36813-5041-488a-a606-b8b3372f8790",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8192,\n",
       " ['A1BG',\n",
       "  'AAAS',\n",
       "  'AACS',\n",
       "  'AAGAB',\n",
       "  'AAK1',\n",
       "  'AAMDC',\n",
       "  'AAMP',\n",
       "  'AAR2',\n",
       "  'AARS',\n",
       "  'AARS2'])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_genes = ds_adata.var_names.tolist()\n",
    "ds_genes = sorted(list(set(gw_genes)))\n",
    "len(ds_genes), ds_genes[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b58fc8f5-d1b7-4355-8113-16e4ff9252fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including all genes\n"
     ]
    }
   ],
   "source": [
    "if len(set(all_genes).union(set(ds_genes))) <= n_genes:\n",
    "    all_genes = set(all_genes).union(set(ds_genes))\n",
    "    print('including all genes')\n",
    "else:\n",
    "    skipped_genes = set(ds_genes) - set(all_genes)\n",
    "    print(f'skipping {skipped_genes}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "11a26286-5296-4535-952a-255348131cef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1543, 1329)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_perts = ds_adata.obs['condition'].unique()\n",
    "ds_perts = [clean_gears_name(p) for p in ds_perts if p != 'ctrl' and p != 'control']\n",
    "ds_perts = sorted(list(set(ds_perts)))\n",
    "remaining_ds_perts = sorted(list(set(ds_perts) - set(hold_out_perts)))\n",
    "len(ds_perts), len(remaining_ds_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c7f27981-4299-4263-a83f-daeee7961854",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9869"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_perts = set(all_perts).union(set(remaining_ds_perts))\n",
    "len(all_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4587ff4f-db72-4cb4-aeb2-50767e637f74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'total cells: 2267989 | total genes 8192 | total perturbations 9869'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'total cells: {total_cells} | total genes {len(all_genes)} | total perturbations {len(all_perts)}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6aaaa9d8-786e-4bcb-a258-8c40db751c13",
   "metadata": {},
   "source": [
    "## adamson Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b46dab77-2f01-421c-b656-656c3be0c9f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_adata = sc.read_h5ad(datasets['adamson'], backed='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "73c98903-145a-408a-93af-331412495d89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(68603, 2336592)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_total_cells = ds_adata.n_obs\n",
    "total_cells += ds_total_cells\n",
    "ds_total_cells, total_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a3601046-f9cf-46f3-9d26-f2c1e5b9f069",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8192,\n",
       " ['A1BG',\n",
       "  'AAAS',\n",
       "  'AACS',\n",
       "  'AAGAB',\n",
       "  'AAK1',\n",
       "  'AAMDC',\n",
       "  'AAMP',\n",
       "  'AAR2',\n",
       "  'AARS',\n",
       "  'AARS2'])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_genes = ds_adata.var_names.tolist()\n",
    "ds_genes = sorted(list(set(gw_genes)))\n",
    "len(ds_genes), ds_genes[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f58ed823-0b95-46ad-8b2a-9b11665b20a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including all genes\n"
     ]
    }
   ],
   "source": [
    "if len(set(all_genes).union(set(ds_genes))) <= n_genes:\n",
    "    all_genes = set(all_genes).union(set(ds_genes))\n",
    "    print('including all genes')\n",
    "else:\n",
    "    skipped_genes = set(ds_genes) - set(all_genes)\n",
    "    print(f'skipping {skipped_genes}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "759f9c19-a38c-45aa-8a79-1ff99dc78c44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(86, 75)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_perts = ds_adata.obs['condition'].unique()\n",
    "ds_perts = [clean_gears_name(p) for p in ds_perts if p != 'ctrl' and p != 'control']\n",
    "ds_perts = sorted(list(set(ds_perts)))\n",
    "remaining_ds_perts = sorted(list(set(ds_perts) - set(hold_out_perts)))\n",
    "len(ds_perts), len(remaining_ds_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "09a8fd19-6f78-4d9f-bd36-78a8f96f20ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9875"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_perts = set(all_perts).union(set(remaining_ds_perts))\n",
    "len(all_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "dfc33d5f-0830-41f5-8f9a-31cc4e005c00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'total cells: 2336592 | total genes 8192 | total perturbations 9875'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'total cells: {total_cells} | total genes {len(all_genes)} | total perturbations {len(all_perts)}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30e8e871-adcd-4bb7-a24c-3413764480a7",
   "metadata": {},
   "source": [
    "## Convert Perts to Embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6c87c091-3315-43d8-981d-e1e23518547e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9875, ['DOT1L', 'NR3C1'])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(all_perts)),list(all_perts)[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a26f19-f6b9-41b6-b1c6-57d6c16ddee1",
   "metadata": {},
   "source": [
    "**load uniprot cache**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "13b5ba2f-6098-4fb3-9a8f-5c367b190eeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/djemec/data/jepa/v0_3/uniprot/UP000005640_9606.fasta.gz')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_prots = 'https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Eukaryota/UP000005640/UP000005640_9606.fasta.gz'\n",
    "#! wget -P {protein_dir} {all_prots}\n",
    "fasta = [i for i in protein_dir.iterdir()][0]\n",
    "fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6bb8e87f-ad27-4a7a-82c6-dfc9927fa34c",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_db = {}\n",
    "with gzip.open(fasta, 'rt') as handle:\n",
    "    for record in SeqIO.parse(handle, 'fasta'):\n",
    "        # Header: >sp|P12345|GENE_HUMAN Description GN=Symbol PE=1 ...\n",
    "        desc = record.description\n",
    "        \n",
    "        # Robustly extract Gene Name (GN)\n",
    "        if 'GN=' in desc:\n",
    "            parts = desc.split('GN=')\n",
    "            if len(parts) > 1:\n",
    "                # Symbol is usually the first string after GN=, before the next space\n",
    "                symbol = parts[1].split(' ')[0]\n",
    "                uniprot_db[symbol] = str(record.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54153ab4-95e2-4b90-853b-45b1bcd6e6ae",
   "metadata": {},
   "source": [
    "**Normalize Gene Perturbation Names**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "99da4076-f073-4f3c-a637-9788704c4af3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "557 input query terms found dup hits:\t[('MEMO1', 2), ('CT45A5', 4), ('SMAD1', 2), ('LIG1', 2), ('LEPR', 2), ('MAGEA6', 2), ('PLD1', 2), ('\n",
      "2 input query terms found no hit:\t['AC015871.1', 'AC118549.1']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(10522,\n",
       " [{'query': 'DOT1L', '_id': '84444', '_score': 17.593918, 'symbol': 'DOT1L'},\n",
       "  {'query': 'NR3C1', '_id': '2908', '_score': 17.616997, 'symbol': 'NR3C1'}])"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "mg_results = mg.querymany(all_perts, scopes='symbol,alias', fields='symbol', species='human')\n",
    "len(mg_results),mg_results[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a4d91136-a19e-4960-bfc3-811b6165c357",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_map = {}\n",
    "missing_genes = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "988635b3-72cb-45d4-9c3e-b8f79fc77eb0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "failed to find {'query': 'AC015871.1', 'notfound': True}\n",
      "failed to find {'query': 'AC118549.1', 'notfound': True}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['HIST1H4B',\n",
       " 'GAS8',\n",
       " 'FUT11',\n",
       " 'B3GNTL1',\n",
       " 'UROS',\n",
       " 'TRMT12',\n",
       " 'C12orf43',\n",
       " 'C18orf21',\n",
       " 'C18orf21',\n",
       " 'AC015871.1',\n",
       " 'ALG1L',\n",
       " 'TMEM241',\n",
       " 'HIST1H2AK',\n",
       " 'TMCO3',\n",
       " 'HBA2',\n",
       " 'AC118549.1',\n",
       " 'RPL17-C18orf32',\n",
       " 'MAGEA9B',\n",
       " 'HIST1H3J',\n",
       " 'ZNF720',\n",
       " 'C11orf58',\n",
       " 'C22orf46',\n",
       " 'MFSD5',\n",
       " 'NEPRO',\n",
       " 'HIST1H3F',\n",
       " 'MFSD3',\n",
       " 'FAM210A',\n",
       " 'HIST1H4H',\n",
       " 'RGPD6',\n",
       " 'MFSD10',\n",
       " 'FAM210B',\n",
       " 'HIST1H2AI',\n",
       " 'MFSD4B',\n",
       " 'HIST1H4I',\n",
       " 'HIST1H4J',\n",
       " 'NME1-NME2',\n",
       " 'CCDC169-SOHLH2',\n",
       " 'HIST1H2AM',\n",
       " 'C1orf56',\n",
       " 'SLC35C2',\n",
       " 'NEDD8-MDP1',\n",
       " 'C11orf54',\n",
       " 'RBAK-RBAKDN',\n",
       " 'HIST1H2AL',\n",
       " 'C21orf91',\n",
       " 'H3F3B',\n",
       " 'FUT10',\n",
       " 'HIST2H4A',\n",
       " 'C1orf43',\n",
       " 'NDUFA4',\n",
       " 'NDUFA4',\n",
       " 'RBM14-RBM4',\n",
       " 'C19orf48',\n",
       " 'ILVBL',\n",
       " 'RPS10-NUDT3',\n",
       " 'ST20-MTHFS',\n",
       " 'HIST1H2AE',\n",
       " 'SERF1B',\n",
       " 'HIST2H3D',\n",
       " 'SMN2',\n",
       " 'HIST1H4D',\n",
       " 'HIST1H3H',\n",
       " 'C5orf15',\n",
       " 'C1orf35',\n",
       " 'EFCAB2',\n",
       " 'HIST1H4E',\n",
       " 'ARPC4-TTLL3',\n",
       " 'KRBA1',\n",
       " 'C10orf88',\n",
       " 'HIST1H4C',\n",
       " 'MFSD9',\n",
       " 'MDM1',\n",
       " 'C2orf88',\n",
       " 'FAM156B']"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for res in mg_results:\n",
    "    original_query = res['query']\n",
    "    \n",
    "    if 'symbol' in res:\n",
    "        modern_symbol = res['symbol']\n",
    "    else:\n",
    "        print(f'failed to find {res}')\n",
    "        modern_symbol = original_query\n",
    "\n",
    "    if modern_symbol in uniprot_db:\n",
    "        final_map[original_query] = uniprot_db[modern_symbol]\n",
    "    else:\n",
    "        missing_genes.append(original_query)\n",
    "\n",
    "# final cleanup of duplicates\n",
    "missing_genes = [g for g in missing_genes if not final_map.get(g)]\n",
    "\n",
    "missing_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e1dc051-9673-41e4-8510-ff18d5f44127",
   "metadata": {},
   "source": [
    "**Search NCBI for missing protein sequences**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c461d969-b2aa-4104-a6a7-7d01065636f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████| 74/74 [01:03<00:00,  1.16it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['AC015871.1', 'ALG1L', 'AC118549.1', 'C22orf46', 'C19orf48']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Entrez.email = 'test@test.com'\n",
    "results = {}\n",
    "for gene in tqdm(missing_genes.copy()):\n",
    "    try:\n",
    "        # Specific search for RefSeq proteins in Humans\n",
    "        term = f'{gene}[All Fields] AND Homo sapiens[Organism] AND srcdb_refseq[PROP]'\n",
    "        \n",
    "        # 1. Search\n",
    "        search_handle = Entrez.esearch(db=\"protein\", term=term, retmax=1)\n",
    "        record = Entrez.read(search_handle)\n",
    "        \n",
    "        if record['IdList']:\n",
    "            # 2. Fetch\n",
    "            uid = record['IdList'][0]\n",
    "            fetch_handle = Entrez.efetch(db=\"protein\", id=uid, rettype=\"fasta\", retmode=\"text\")\n",
    "            seq_record = SeqIO.read(fetch_handle, \"fasta\")\n",
    "            results[gene] = str(seq_record.seq)\n",
    "            missing_genes.remove(gene)\n",
    "        else:\n",
    "            results[gene] = \"M\" # True Ghost Gene\n",
    "    except Exception as e:\n",
    "        print(f'Failed Entrez lookup for {gene}: {e}')\n",
    "        results[gene] = \"M\"\n",
    "        \n",
    "final_map.update(results)\n",
    "\n",
    "missing_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae9d1a4e-1834-4731-bbbe-90f5d1bbe216",
   "metadata": {},
   "source": [
    "**Manually handle unmapped**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "dbf3b0e5-26ed-45f7-8fc3-a6bc06c55e4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████| 5/5 [00:03<00:00,  1.27it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clone_map = {\n",
    "    'AC118549.1': 'ZZZ3',\n",
    "    'AC015871.1': 'LINC01587', # Will fail protein search (Correctly)\n",
    "}\n",
    "results = {}\n",
    "for gene in tqdm(missing_genes.copy()):\n",
    "    \n",
    "    term = clone_map.get(gene, gene)\n",
    "    try:\n",
    "        # Specific search for RefSeq proteins in Humans\n",
    "        term = f'{term}[All Fields] AND Homo sapiens[Organism]'\n",
    "        \n",
    "        # 1. Search\n",
    "        search_handle = Entrez.esearch(db=\"protein\", term=term, retmax=1)\n",
    "        record = Entrez.read(search_handle)\n",
    "        \n",
    "        if record['IdList']:\n",
    "            # 2. Fetch\n",
    "            uid = record['IdList'][0]\n",
    "            fetch_handle = Entrez.efetch(db=\"protein\", id=uid, rettype=\"fasta\", retmode=\"text\")\n",
    "            seq_record = SeqIO.read(fetch_handle, \"fasta\")\n",
    "            results[gene] = str(seq_record.seq)\n",
    "            missing_genes.remove(gene)\n",
    "        else:\n",
    "            results[gene] = \"M\" # True Ghost Gene\n",
    "    except Exception as e:\n",
    "        print(f'Failed Entrez lookup for {gene}: {e}')\n",
    "        results[gene] = \"M\"\n",
    "        \n",
    "final_map.update(results)\n",
    "\n",
    "missing_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e06f0d6-ac1c-4702-9a13-44a9e6b5a5fd",
   "metadata": {},
   "source": [
    "**remove final missing pieces for now**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "af113872-c968-44dd-8aeb-7510f35b5089",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9875"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_perts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "552e3dcb-065f-4805-9fc1-16a0bc4e1195",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9875, 9875)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_perts -= set(missing_genes)\n",
    "\n",
    "for gene in missing_genes:\n",
    "    del final_map[gene]\n",
    "\n",
    "len(all_perts), len(final_map.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "42510baf-acb7-4f93-a357-083c304316dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check that we have protein sequences\n",
    "[g for g in final_map.keys() if final_map[g] == 'M']\n",
    "# check all hold_out perts still encoded\n",
    "[i for i in list(hold_out_perts) if i not in all_perts]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "546eeed2-263e-4d84-a64f-28f92446b985",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pert_dir /'perturbation_seq.json', 'w') as f:\n",
    "    json.dump({str(k): str(v) for k, v in final_map.items()}, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23ee56de-5311-4699-8c78-6c60e6e04479",
   "metadata": {},
   "source": [
    "**Get Model For Embeddings**\n",
    "for protein embeddings, we'll use the [ESM-2](https://huggingface.co/facebook/esm2_t6_8M_UR50D) for now given it's been a common workhorse. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "fd7ee334-766e-4e33-ae7f-d1ced2bf1202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "using cpu\n"
     ]
    }
   ],
   "source": [
    "model_name = 'facebook/esm2_t6_8M_UR50D' \n",
    "def get_device():\n",
    "    device = 'cpu'\n",
    "    if torch.cuda.is_available():\n",
    "        torch.cuda.manual_seed(1337)\n",
    "        device = 'cuda'\n",
    "    print(f'using {device}')\n",
    "    return device\n",
    "\n",
    "DEVICE = get_device()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1f62b1f4-a053-4162-acbb-5da08f11cb19",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of EsmModel were not initialized from the model checkpoint at facebook/esm2_t6_8M_UR50D and are newly initialized: ['pooler.dense.bias', 'pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    }
   ],
   "source": [
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "model = AutoModel.from_pretrained(model_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "370414cf-318f-4453-b175-f0c093d62059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EsmModel(\n",
       "  (embeddings): EsmEmbeddings(\n",
       "    (word_embeddings): Embedding(33, 320, padding_idx=1)\n",
       "    (dropout): Dropout(p=0.0, inplace=False)\n",
       "  )\n",
       "  (encoder): EsmEncoder(\n",
       "    (layer): ModuleList(\n",
       "      (0-5): 6 x EsmLayer(\n",
       "        (attention): EsmAttention(\n",
       "          (self): EsmSelfAttention(\n",
       "            (query): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (key): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (value): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (rotary_embeddings): RotaryEmbedding()\n",
       "          )\n",
       "          (output): EsmSelfOutput(\n",
       "            (dense): Linear(in_features=320, out_features=320, bias=True)\n",
       "            (dropout): Dropout(p=0.0, inplace=False)\n",
       "          )\n",
       "          (LayerNorm): LayerNorm((320,), eps=1e-05, elementwise_affine=True)\n",
       "        )\n",
       "        (intermediate): EsmIntermediate(\n",
       "          (dense): Linear(in_features=320, out_features=1280, bias=True)\n",
       "        )\n",
       "        (output): EsmOutput(\n",
       "          (dense): Linear(in_features=1280, out_features=320, bias=True)\n",
       "          (dropout): Dropout(p=0.0, inplace=False)\n",
       "        )\n",
       "        (LayerNorm): LayerNorm((320,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "    )\n",
       "    (emb_layer_norm_after): LayerNorm((320,), eps=1e-05, elementwise_affine=True)\n",
       "  )\n",
       "  (pooler): EsmPooler(\n",
       "    (dense): Linear(in_features=320, out_features=320, bias=True)\n",
       "    (activation): Tanh()\n",
       "  )\n",
       "  (contact_head): EsmContactPredictionHead(\n",
       "    (regression): Linear(in_features=120, out_features=1, bias=True)\n",
       "    (activation): Sigmoid()\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.eval()\n",
    "model.to(DEVICE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a8724d5-d021-4f4e-8af5-4a2e809de435",
   "metadata": {},
   "source": [
    "**Get Embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "9293d339-82e0-4f1c-b15f-e366739d88c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_genes = sorted(list(final_map.keys()))\n",
    "    \n",
    "embeddings = []\n",
    "pert_to_id = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "89a0596f-d76a-4a35-a514-440ac0ce8fe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████| 9875/9875 [05:42<00:00, 28.82it/s]\n"
     ]
    }
   ],
   "source": [
    "with torch.no_grad():\n",
    "    for idx, gene in enumerate(tqdm(sorted_genes)):\n",
    "        seq = final_map[gene]\n",
    "        \n",
    "        # Map Gene Name -> Integer ID\n",
    "        pert_to_id[gene] = idx\n",
    "        \n",
    "        # Tokenize\n",
    "        # Truncate to 1024 AA (covers 95% of human proteins)\n",
    "        inputs = tokenizer(seq, return_tensors=\"pt\", truncation=True, max_length=1024)\n",
    "        inputs = {k: v.to(DEVICE) for k, v in inputs.items()}\n",
    "        \n",
    "        # Forward Pass\n",
    "        outputs = model(**inputs)\n",
    "        \n",
    "        # Mean Pooling: Compress [Seq_Len, 320] -> [320]\n",
    "        # We take the mean of all amino acids to get the \"whole protein\" vector\n",
    "        phys_vector = outputs.last_hidden_state[0].mean(dim=0).cpu().numpy()\n",
    "        embeddings.append(phys_vector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6af10658-47c2-46ec-a039-879d40005005",
   "metadata": {},
   "outputs": [],
   "source": [
    "pert_to_id['control'] = len(embeddings)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a61ff314-7ce9-4ea1-adcb-3bf9e2bdbae7",
   "metadata": {},
   "source": [
    "**Save Pert Mapping/Embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "758e58f3-0589-4d2b-80ce-36ac6749ddaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_bank = np.vstack(embeddings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3950590c-98a1-4b30-bb56-50639f68b3d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(pert_dir / 'action_embeddings_esm2.npy', final_bank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "785ca0bc-353b-4aa1-9435-fd0616c6d302",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(pert_dir / 'pert_to_id.json', 'w') as f:\n",
    "        json.dump(pert_to_id, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fe0428d-a676-458a-acd1-c3614f85ffab",
   "metadata": {},
   "source": [
    "## Save Genes and perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b571fc68-1c4e-42a3-920f-c3a6811c999b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_and_perts = {\n",
    "    'all_genes':list(all_genes),\n",
    "    'anchor_genes': list(anchor_genes),\n",
    "    'all_perts': list(all_perts),\n",
    "    'hold_out_perts': list(hold_out_perts)\n",
    "}\n",
    "\n",
    "with open(raw_dir / 'gene_and_perts.json', 'w') as f:\n",
    "    json.dump(genes_and_perts, f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
